File: computer3/Fall Term/Solutions/Class08_Oct_30/lvl3_giant_ants.cpp
#include <bits/stdc++.h>
using namespace std;
int N, M, W, ant[105], dis[105]; vector<int> adj[105]; bool vis[105]; queue<int> Q;
int main(){
    cin >> N >> M;
    for(int i=0, x, y; i<M; i++){
        cin >> x >> y;
        adj[x].push_back(y); adj[y].push_back(x);
    }
    cin >> W;
    for(int i=0, x; i<W; i++){
        cin >> x;
        Q.push(x); ant[x]=0; vis[x]=1;
    }
    while(!Q.empty()){
        int cur = Q.front(); Q.pop();
        for(int nxt: adj[cur]){
            if(!vis[nxt]){
                Q.push(nxt); ant[nxt]=ant[cur]+4; vis[nxt]=1;
            }
        }
    }
    memset(vis, 0, sizeof(vis));
    Q.push(1); dis[1]=0; vis[1]=1;
    while(!Q.empty()){
        int cur = Q.front(); Q.pop();
        for(int nxt: adj[cur]){
            if(!vis[nxt] && dis[cur]+1 < ant[nxt]){
                dis[nxt]=dis[cur]+1; vis[nxt]=1; Q.push(nxt);
            }
        }
    }
    if(vis[N]) cout << dis[N] << endl;
    else cout << "sacrifice bobhob314" << endl;
}

File: computer3/Fall Term/Solutions/Class08_Oct_30/lvl3_giant_ants.java
import java.util.*;
import java.io.*;
public class Main {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static PrintWriter pr = new PrintWriter(new BufferedWriter(new OutputStreamWriter(System.out)));
	static StringTokenizer st;
	public static void main(String[] args) throws IOException{
		int n = readInt(), m = readInt(); List<Integer> adj[] = new ArrayList[n+1];
		for(int i=1; i<=n; i++) adj[i] = new ArrayList<>();
		for(int i=1; i<=m; i++) {
			int u = readInt(), v = readInt();
			adj[u].add(v); adj[v].add(u);
		}
		int w = readInt(), dis[] = new int[n+1]; //run Ants BFS
		Queue<Integer> q = new LinkedList<>();  boolean vis[] = new boolean[n+1];
		for(int i=1; i<=w; i++) {
			int x = readInt(); q.add(x); vis[x]=true; 
		}
		while(!q.isEmpty()) {
			int cur = q.poll();
			for(int nxt : adj[cur]) {
				if(!vis[nxt]) { q.add(nxt); vis[nxt]=true; dis[nxt] = dis[cur] + 4; }
			}
		}
		Arrays.fill(vis, false);
		int human[] = new int[n+1];
		q.add(1); vis[1] = true;  
		while(!q.isEmpty()) {
			int cur = q.poll();
			for(int nxt : adj[cur]) {
				if(!vis[nxt] && human[cur] + 1 < dis[nxt]) {
					q.add(nxt); vis[nxt]=true; human[nxt]=human[cur] + 1;
				}
			}
		}
		if(vis[n]) System.out.println(human[n]);
		else System.out.println("sacrifice bobhob314");
	}
	static String next () throws IOException {
		while (st == null || !st.hasMoreTokens())
			st = new StringTokenizer(br.readLine().trim());
		return st.nextToken();
	}
	static long readLong () throws IOException {
		return Long.parseLong(next());
	}
	static int readInt () throws IOException {
		return Integer.parseInt(next());
	}
	static double readDouble () throws IOException {
		return Double.parseDouble(next());
	}
	static char readCharacter () throws IOException {
		return next().charAt(0);
	}
	static String readLine () throws IOException {
		return br.readLine().trim();
	}
}